src/widgets/RealtimeEffectPanel.cpp

```diff
@@ -0,0 +1,813 @@
+               XO("'%s' effects reorder %d->%d").Format(mTrack->GetName(), from, to),
+               //i18n-hint: undo history record
+               XO("Realtime effect reorder"), UndoPush::CONSOLIDATE);
+         }
+         else
+         {
+            wxWindowUpdateLocker freeze(this);
+            Layout();
+         }
+      });
+      SetScrollRate(0, 20);
+   }
+
+   void OnEffectListItemChange(const RealtimeEffectListMessage& msg)
+   {
+      wxWindowUpdateLocker freeze(this);
+      if(msg.type == RealtimeEffectListMessage::Type::Move)
+      {
+         const auto sizer = mEffectListContainer->GetSizer();
+
+         const auto movedItem = sizer->GetItem(msg.index);
+
+         const auto proportion = movedItem->GetProportion();
+         const auto flag = movedItem->GetFlag();
+         const auto border = movedItem->GetBorder();
+
+         auto window = movedItem->GetWindow();
+         
+         sizer->Remove(msg.index);
+         sizer->Insert(std::any_cast<size_t>(msg.extra), window, proportion, flag, border);
+      }
+      else if(msg.type == RealtimeEffectListMessage::Type::Insert)
+      {
+         auto& effects = RealtimeEffectList::Get(*mTrack);
+         InsertEffectRow(msg.index, effects.GetStateAt(msg.index));
+      }
+      else if(msg.type == RealtimeEffectListMessage::Type::Remove)
+      {
+         auto sizer = mEffectListContainer->GetSizer();
+         auto item = sizer->GetItem(msg.index)->GetWindow();
+         item->Destroy();
+      }
+      Layout();
+   }
+
+   void ResetTrack()
+   {
+      mEffectListItemMovedSubscription.Reset();
+      mUndoManagerSubscription.Reset();
+
+      mTrack.reset();
+      ReloadEffectsList();
+   }
+
+   void SetTrack(AudacityProject& project, const std::shared_ptr<Track>& track)
+   {
+      if(mTrack == track)
+         return;
+
+      mEffectListItemMovedSubscription.Reset();
+      mUndoManagerSubscription.Reset();
+      
+      mTrack = track;
+      ReloadEffectsList();
+
+      if(track)
+      {
+         auto& effects = RealtimeEffectList::Get(*mTrack);
+         mEffectListItemMovedSubscription = effects.Subscribe(*this, &RealtimeEffectListWindow::OnEffectListItemChange);
+         mUndoManagerSubscription = UndoManager::Get(project).Subscribe(*this, &RealtimeEffectListWindow::OnUndoRedoMessage);
+      }
+   }
+
+   void ReloadEffectsList()
+   {
+      wxWindowUpdateLocker freeze(this);
+      //delete only layouted items
+      mEffectListContainer->GetSizer()->Clear(true);
+      if(mTrack)
+      {
+         auto& effects = RealtimeEffectList::Get(*mTrack);
+         for(size_t i = 0, count = effects.GetStatesCount(); i < count; ++i)
+            InsertEffectRow(i, effects.GetStateAt(i));
+      }
+      mAddEffect->Enable(!!mTrack);
+      Layout();
+   }
+
+   void OnUndoRedoMessage(const UndoRedoMessage& msg)
+   {
+      if(msg.type == UndoRedoMessage::UndoOrRedo || msg.type == UndoRedoMessage::Reset)
+         ReloadEffectsList();
```

Comment:
I tested this, and Undo/Redo did not have the desired effect, and I figured out why.

`mTrack` is a shared pointer, so there isn't a danger of a dangling pointer.  However, it must be reassigned when the undo state is pushed or popped, so that it points at a track in the current version of the track list.

See `Track::GetId()` for an identifier of a track across undo/redo states.  This window needs to store that ID, then find that ID again with each undo and redo.

It may also be necessary to `Refresh()` after `SetTrack()`.

Reply:
Thanks!

Reply:
But turns out this is not neccessary. `ProjectWindow` gets notification from `TrackFocus` about changed focus, and given that focus now points to a new track instance `RealtimeEffectPanel` updates anyway